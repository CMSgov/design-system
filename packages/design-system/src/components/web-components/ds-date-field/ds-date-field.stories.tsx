import { useEffect } from 'react';
import { action } from '@storybook/addon-actions';
import type { Meta } from '@storybook/react';
import WebComponentDocTemplate from '../../../../../../.storybook/docs/WebComponentDocTemplate.mdx';
import { webComponentDecorator } from '../storybook';
import './ds-date-field';

const formatDateArg = (dateArg: string | number | undefined): string => {
  if (typeof dateArg === 'number') {
    return new Date(dateArg).toISOString();
  }
  return dateArg;
};

const meta: Meta = {
  title: 'Web Components/ds-date-field',
  decorators: [webComponentDecorator],
  parameters: {
    docs: {
      page: WebComponentDocTemplate,
    },
  },
  args: {
    hint: 'If you were born on a leap day, entering the date will either crash our servers or open a portal to an alternate dimension.',
    label: 'Enter your date of birth.',
    name: 'ds-date-field',
  },
  argTypes: {
    name: {
      description: "The `input` field's `name` attribute.",
      control: 'text',
    },
    'root-id': {
      description:
        "A unique ID for this element. A unique ID will be generated if one isn't provided.",
      control: 'text',
    },
    inversed: {
      description: 'Set to `true` to apply the "inverse" color scheme.',
      control: 'boolean',
    },
    value: {
      description:
        "The `input` field's `value` attribute. Typically not used for uncontrolled components (web components).",
      control: 'text',
    },
    'default-value': {
      description: `Sets the initial value for the component. This value will be used until the user changes the input, at which point the new value will be handled by the component's internal state.
                    The \`onChange\` event will still be triggered with the updated value, but the component will manage the state internally.`,
      control: 'text',
    },
    'default-month': {
      description:
        'The initial month to show in the calendar picker. If not provided, defaults to the month of the currently selected date.',
      control: 'date',
    },
    'from-date': {
      description:
        'Earliest day to start month navigation in the calendar picker. (This does not restrict dates entered manually.)',
      control: 'date',
    },
    'from-month': {
      description:
        'Earliest month to start month navigation in the calendar picker. (This does not restrict dates entered manually.)',
      control: 'date',
    },
    'from-year': {
      description:
        'Earliest year to start month navigation in the calendar picker. (This does not restrict dates entered manually.)',
      control: 'text',
    },
    'to-date': {
      description:
        'Latest day to end month navigation in the calendar picker. (This does not restrict dates entered manually.)',
      control: 'date',
    },
    'to-month': {
      description:
        'Latest month to end month navigation in the calendar picker. (This does not restrict dates entered manually.)',
      control: 'date',
    },
    'to-year': {
      description:
        'Latest year to end month navigation in the calendar picker. (This does not restrict dates entered manually.)',
      control: 'text',
    },
    hint: {
      description: 'Hint text or HTML.',
      control: 'text',
    },
    'hint-class-name': {
      control: 'text',
      description: 'Custom CSS class name(s) for the hint element',
    },
    'hint-id': {
      description: 'The ID of the hint element',
      control: 'text',
    },
    label: {
      description: 'Label text or HTML.',
      control: 'text',
    },
    'label-class-name': {
      description: 'Additional classes to be added to the field label',
      control: 'text',
    },
    'label-id': {
      description:
        "A unique `id` to be used on the field label. If one isn't provided, a unique ID will be generated.",
      control: 'text',
    },
    'requirement-label': {
      description:
        'Text showing the requirement (e.g., "Optional", or "Required"). In most cases, this should be used to indicate which fields are optional. See the [form guidelines](https://design.cms.gov/patterns/Forms/forms/) for more info.',
      control: 'text',
    },
    'error-message': {
      description: 'Enable the error state by providing an error message.',
      control: 'text',
    },
    'error-id': {
      description:
        'The ID of the error message applied to this field. If none is provided, the id will be derived from the `root-id` attribute.',
      control: 'text',
    },
    'error-placement': {
      description: 'Location of the error message relative to the field input',
      options: ['top', 'bottom'],
      control: { type: 'radio' },
    },
  },
};
export default meta;

const Template = (args) => {
  useEffect(() => {
    const element = document.querySelector('ds-date-field');
    if (element) {
      const handleStoryBookChange = (
        event: CustomEvent<{ updatedValue: string; formattedValue: string }>
      ) => {
        action('ds-change')(event);
      };
      const handleStoryBookBlur = (event: Event) => {
        action('ds-blur')(event);
      };
      element.addEventListener('ds-change', handleStoryBookChange as EventListener);
      element.addEventListener('ds-blur', handleStoryBookBlur as EventListener);
      return () => {
        element.removeEventListener('ds-change', handleStoryBookChange as EventListener);
        element.removeEventListener('ds-blur', handleStoryBookBlur as EventListener);
      };
    }
  }, []);

  const formattedArgs = {
    ...args,
    'to-date': formatDateArg(args['to-date']),
    'to-month': formatDateArg(args['to-month']),
    'from-month': formatDateArg(args['from-month']),
    'default-month': formatDateArg(args['default-month']),
    'from-date': formatDateArg(args['from-date']),
  };

  return <ds-date-field {...formattedArgs} />;
};

export const Default = Template.bind({});
Default.args = {
  name: 'ds-date-field',
};

export const WithPicker = Template.bind({});
WithPicker.args = {
  label: 'What day did you move?',
  hint: 'This date should be within the past 60 days in order to qualify.',
  'from-year': '2023',
  'to-date': new Date('2023-02-15T21:56:34.272Z').toISOString(),
};

export const WithError = Template.bind({});
WithError.args = {
  'error-message': 'This is an example error message.',
  ...WithPicker.args,
};
