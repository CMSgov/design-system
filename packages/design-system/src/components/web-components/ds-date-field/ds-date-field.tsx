import { define } from '../preactement/define';
import { SingleInputDateFieldProps } from '../../DateField/SingleInputDateField';
import { SingleInputDateField } from '../../DateField';
import { parseBooleanAttr, parseDateAttr, parseIntegerAttr } from '../wrapperUtils';
import { formAttrs } from '../shared-attributes/form';

const attributes = [
  'name',
  'root-id',
  'inversed',
  'value',
  'default-month',
  'from-date',
  'from-month',
  'from-year',
  'to-date',
  'to-month',
  'to-year',
  ...formAttrs,
] as const;

type IncompatibleProps =
  | 'inversed'
  | 'defaultMonth'
  | 'fromDate'
  | 'fromMonth'
  | 'fromYear'
  | 'toDate'
  | 'toDate'
  | 'toMonth'
  | 'toYear';

interface WrapperProps extends Omit<SingleInputDateFieldProps, IncompatibleProps> {
  rootId: string;
  inversed?: string;
  defaultMonth?: string;
  fromDate?: string;
  fromMonth?: string;
  fromYear?: string;
  toDate?: string;
  toMonth?: string;
  toYear?: string;
}

const Wrapper = ({
  rootId,
  inversed,
  defaultMonth,
  fromDate,
  fromMonth,
  fromYear,
  toDate,
  toMonth,
  toYear,
  value,
  ...otherProps
}: WrapperProps) => {
  return (
    <SingleInputDateField
      id={rootId}
      inversed={parseBooleanAttr(inversed)}
      defaultValue={value}
      defaultMonth={parseDateAttr(defaultMonth)}
      fromDate={parseDateAttr(fromDate)}
      fromMonth={parseDateAttr(fromMonth)}
      fromYear={parseIntegerAttr(fromYear)}
      toDate={parseDateAttr(toDate)}
      toMonth={parseDateAttr(toMonth)}
      toYear={parseIntegerAttr(toYear)}
      {...otherProps}
    />
  );
};

/* eslint-disable @typescript-eslint/no-namespace */
declare global {
  namespace JSX {
    interface IntrinsicElements {
      'ds-date-field': JSX.IntrinsicElements['div'] & {
        [K in (typeof attributes)[number]]?: string;
      };
    }
  }
}
/* eslint-enable */

define('ds-date-field', () => Wrapper, {
  attributes,
  events: [
    [
      'onChange',
      (updatedValue: string, formattedValue: string) => ({
        detail: { updatedValue, formattedValue },
      }),
    ],
    'onBlur',
  ],
} as any);
