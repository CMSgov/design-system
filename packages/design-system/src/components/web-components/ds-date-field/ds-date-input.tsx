import { define } from '../preactement/define';
import { DateInput, DateInputProps } from '../../DateField/DateInput';

const attributes = [
  'class-name',
  'auto-complete',
  'disabled',
  'inversed',
  'id',
  'day-label',
  'day-name',
  'day-default-value',
  'day-invalid',
  'day-value',
  'month-label',
  'month-name',
  'month-default-value',
  'month-invalid',
  'month-value',
  'year-label',
  'year-name',
  'year-default-value',
  'year-invalid',
  'year-value',
] as const;

interface WrapperProps extends DateInputProps {
  className?: string;
  autoComplete?: boolean;
  disabled?: boolean;
  inversed?: boolean;
  id?: string;
  dayLabel?: string;
  dayName: string;
  dayDefaultValue?: string | number;
  dayInvalid?: boolean;
  dayValue?: string | number;
  monthLabel?: string;
  monthName: string;
  monthDefaultValue?: string | number;
  monthInvalid?: boolean;
  monthValue?: string | number;
  yearLabel?: string;
  yearName: string;
  yearDefaultValue?: string | number;
  yearInvalid?: boolean;
  yearValue?: string | number;
}

const Wrapper = ({
  className,
  autoComplete,
  disabled,
  inversed,
  id,
  dayLabel,
  dayName,
  dayDefaultValue,
  dayInvalid,
  dayValue,
  monthLabel,
  monthName,
  monthDefaultValue,
  monthInvalid,
  monthValue,
  yearLabel,
  yearName,
  yearDefaultValue,
  yearInvalid,
  yearValue,
  ...otherProps
}: WrapperProps) => {
  return (
    <DateInput
      className={className}
      autoComplete={autoComplete}
      disabled={disabled}
      inversed={inversed}
      id={id}
      dayLabel={dayLabel}
      dayName={dayName}
      dayDefaultValue={dayDefaultValue}
      dayInvalid={dayInvalid}
      dayValue={dayValue}
      monthLabel={monthLabel}
      monthName={monthName}
      monthDefaultValue={monthDefaultValue}
      monthInvalid={monthInvalid}
      monthValue={monthValue}
      yearLabel={yearLabel}
      yearName={yearName}
      yearDefaultValue={yearDefaultValue}
      yearInvalid={yearInvalid}
      yearValue={yearValue}
      {...otherProps}
    />
  );
};

/* eslint-disable @typescript-eslint/no-namespace */
declare global {
  namespace JSX {
    interface IntrinsicElements {
      'ds-date-input': JSX.IntrinsicElements['div'] & {
        [K in (typeof attributes)[number]]?: string;
      };
    }
  }
}
/* eslint-enable */

define('ds-date-input', () => Wrapper, { attributes, events: ['onChange', 'onBlur'] } as any);
// define('ds-date-input', () => DateInput, { attributes, events: ['onChange', 'onBlur'] } as any);
