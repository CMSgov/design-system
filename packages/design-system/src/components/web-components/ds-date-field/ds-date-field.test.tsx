import { render, screen } from '@testing-library/preact';
import userEvent from '@testing-library/user-event';
import './ds-date-field';

const defaultProps = {
  label: 'Birthday',
  hint: 'Please enter your birthday',
  name: 'ds-date-field',
  value: '',
  'root-id': 'static-id',
  // 'ds-change': jest.fn(),
};

function renderField(props = {}) {
  return {
    user: userEvent.setup(),
    ...render(<ds-date-field {...props} />),
  };
}

function renderPicker(props = {}) {
  return renderField({ ...defaultProps, ...props });
}

function renderError(props = {}) {
  return renderField({ ...defaultProps, ...props });
}

function getInput() {
  return screen.getByRole('textbox');
}

describe('DateField', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    document.body.innerHTML = '';
  });
  it('renders without picker', () => {
    renderField(defaultProps);

    const label = screen.getByText('Birthday');
    expect(label).toBeInTheDocument();

    const hint = screen.getByText('Please enter your birthday');
    expect(hint).toBeInTheDocument();

    const masks = screen.getAllByText('MM/DD/YYYY');
    const mask = masks[0];
    expect(mask).toBeInTheDocument();

    const input = getInput();
    expect(input).toMatchSnapshot();
    expect(input).toBeInTheDocument();
    expect(input).toHaveAttribute('type', 'text');
    expect(input).toHaveAttribute('inputmode', 'numeric');
    expect(input).toHaveAttribute('aria-describedby');
    expect(input).toHaveAttribute('aria-invalid', 'false');

    expect(screen.queryByRole('button')).not.toBeInTheDocument();
  });
  it('masks in label', async () => {
    const { user } = renderField({ value: '11-01' });

    await user.click(getInput());

    const mask = screen.getByText(/11\/01\/YYYY/, {
      selector: 'span:not(.ds-u-visibility--screen-reader)',
    });
    expect(mask).toBeInTheDocument();
  });

  it('generates ids when no id is provided', () => {
    // Render component without 'root-id' prop.
    renderField({
      label: 'Birthday',
      hint: 'Please enter your birthday',
      name: 'ds-date-field',
      value: '',
    });
    const inputElement = getInput();
    expect(inputElement).toMatchSnapshot();
    expect(inputElement.id).toMatch(/^date-field--[\w:.-]+$/);
  });

  it('calls ds-change (onChange) when input changes', async () => {
    const { user } = renderField(defaultProps);

    const dateFieldElement = document.querySelector('ds-date-field');
    const mockChangeHandler = jest.fn();

    dateFieldElement.addEventListener('ds-change', mockChangeHandler);

    const input = getInput() as HTMLInputElement;
    await user.type(input, '1');

    expect(mockChangeHandler).toHaveBeenCalledTimes(1);
    expect(input.value).toBe('1');
    expect(mockChangeHandler.mock.lastCall[0].detail).toEqual({
      updatedValue: '1',
      formattedValue: '01',
    });

    dateFieldElement.removeEventListener('ds-change', mockChangeHandler);
  });

  it('calls ds-change and ds-blur when input loses focus', async () => {
    const { user } = renderField({ value: '01-02-2000' });

    const dateFieldElement = document.querySelector('ds-date-field');
    const mockChangeHandler = jest.fn();
    const mockBlurHandler = jest.fn();
    dateFieldElement.addEventListener('ds-change', mockChangeHandler);
    dateFieldElement.addEventListener('ds-blur', mockBlurHandler);

    const input = getInput() as HTMLInputElement;

    await user.click(input);
    expect(input).toHaveFocus();

    await user.tab();
    expect(input).not.toHaveFocus();
    expect(mockBlurHandler).toHaveBeenCalledTimes(1);
    expect(mockChangeHandler).toHaveBeenCalledTimes(1);
    expect(mockChangeHandler.mock.lastCall[0].detail).toEqual({
      updatedValue: '01/02/2000',
      formattedValue: '01/02/2000',
    });

    dateFieldElement.removeEventListener('ds-change', mockChangeHandler);
    dateFieldElement.removeEventListener('ds-blur', mockBlurHandler);
  });

  describe('DateField with picker', () => {
    const defaultPickerProps = {
      label: 'What day did you move?',
      hint: 'This date should be within the past 60 days in order to qualify',
      'from-year': new Date('01-02-2000').getFullYear(),
      'to-date': new Date('01-02-2000').toISOString().split('T')[0],
    };

    it('renders with picker', async () => {
      const { user } = renderPicker(defaultPickerProps);
      const label = document.querySelector('.ds-c-label');
      const hint = document.querySelector('.ds-c-hint');
      expect(document.querySelector('.ds-c-single-input-date-field')).toBeInTheDocument();
      expect(document.querySelector('.ds-c-label')).toBeInTheDocument();
      expect(document.querySelector('.ds-c-hint')).toBeInTheDocument();

      const wrapper = document.querySelector('.ds-c-single-input-date-field__field-wrapper');
      expect(wrapper).toBeInTheDocument();
      expect(wrapper.querySelector('.ds-c-field')).toBeInTheDocument();

      const button = screen.getByRole('button');
      expect(button).toMatchSnapshot();
      expect(button).toBeInTheDocument();
      expect(button).toHaveClass('ds-c-single-input-date-field__button');
      expect(button).toHaveAttribute('aria-describedby', `${label.id} ${hint.id}`);
      expect(button.firstElementChild.tagName).toBe('svg');
      expect(button.firstElementChild.classList).toContain('ds-c-icon--calendar');

      await user.click(button);
      expect(screen.getByRole('dialog')).toBeInTheDocument();
    });
  });
  describe('Date Field with Error', () => {
    const defaultErrorProps = {
      label: 'What day did you move?',
      hint: 'This date should be within the past 60 days in order to qualify.',
      'from-year': '2023',
      'to-date': new Date('2023-02-15T21:56:34.272Z').toISOString(),
      'error-message': 'This is an example error message.',
    };
    it('renders with error message', () => {
      renderError(defaultErrorProps);
      const errorMessage = screen.getByText('This is an example error message.');
      expect(errorMessage).toBeInTheDocument();
      expect(errorMessage).toMatchSnapshot();
      const inputElement = getInput();
      expect(inputElement).toMatchSnapshot();
      expect(inputElement).toHaveAttribute('aria-invalid', 'true');
      expect(errorMessage).toHaveAttribute('id', expect.stringContaining('static-id__error'));
      expect(inputElement).toHaveAttribute(
        'aria-describedby',
        expect.stringContaining(errorMessage.id)
      );
    });
  });
});
