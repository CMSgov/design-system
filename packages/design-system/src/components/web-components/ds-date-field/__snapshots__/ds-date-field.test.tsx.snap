// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateField Date Field with Error renders with error message 1`] = `
<p
  aria-atomic="true"
  aria-live="assertive"
  class="ds-c-inline-error"
  id="static-id__error"
>
  <svg
    aria-hidden="true"
    class="ds-c-icon ds-c-icon--alert-circle "
    viewBox="36 -12 186 186"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M162.18 41.592c-5.595-9.586-13.185-17.176-22.771-22.771-9.588-5.595-20.055-8.392-31.408-8.392-11.352 0-21.821 2.797-31.408 8.392-9.587 5.594-17.177 13.184-22.772 22.771-5.596 9.587-8.393 20.057-8.393 31.408 0 11.351 2.798 21.82 8.392 31.408 5.595 9.584 13.185 17.176 22.772 22.771 9.587 5.595 20.056 8.392 31.408 8.392s21.822-2.797 31.408-8.392c9.586-5.594 17.176-13.185 22.771-22.771C167.773 94.82 170.57 84.35 170.57 73c0-11.351-2.797-21.822-8.39-31.408zm-43.75 70.433c0 .761-.246 1.398-.734 1.914s-1.086.773-1.793.773H100.26c-.706 0-1.331-.271-1.874-.814-.543-.543-.814-1.168-.814-1.873V96.546c0-.706.271-1.331.814-1.874.543-.543 1.168-.814 1.874-.814h15.643c.707 0 1.306.258 1.793.773.488.518.734 1.154.734 1.915v15.479zm-.164-28.026c-.055.543-.339 1.019-.854 1.426-.517.407-1.154.61-1.914.61h-15.073c-.761 0-1.413-.203-1.956-.61-.543-.407-.815-.883-.815-1.426l-1.385-50.595c0-.653.271-1.141.814-1.467.544-.434 1.196-.652 1.956-.652h17.926c.761 0 1.412.217 1.955.652.543.326.813.815.813 1.467l-1.467 50.595z"
    />
  </svg>
  <span
    class="ds-u-visibility--screen-reader"
  >
    Error: 
  </span>
  This is an example error message.
</p>
`;

exports[`DateField Date Field with Error renders with error message 2`] = `
<input
  aria-describedby="static-id__error static-id__hint static-id__label-mask"
  aria-invalid="true"
  class="ds-c-field ds-c-field--error"
  id="static-id"
  inputmode="numeric"
  name="ds-date-field"
  type="text"
/>
`;

exports[`DateField DateField with picker renders with picker 1`] = `
<button
  aria-describedby="static-id__label static-id__hint"
  class="ds-c-single-input-date-field__button"
  type="button"
>
  <svg
    aria-hidden="false"
    aria-labelledby="static-id__icon__title"
    class="ds-c-icon ds-c-icon--calendar "
    id="static-id__icon"
    role="img"
    viewBox="0 0 448 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title
      id="static-id__icon__title"
    >
      Open calendar picker dialog
    </title>
    <path
      d="M152 64H296V24C296 10.75 306.7 0 320 0C333.3 0 344 10.75 344 24V64H384C419.3 64 448 92.65 448 128V448C448 483.3 419.3 512 384 512H64C28.65 512 0 483.3 0 448V128C0 92.65 28.65 64 64 64H104V24C104 10.75 114.7 0 128 0C141.3 0 152 10.75 152 24V64zM48 448C48 456.8 55.16 464 64 464H384C392.8 464 400 456.8 400 448V192H48V448z"
    />
  </svg>
</button>
`;

exports[`DateField generates ids when no id is provided 1`] = `
<input
  aria-describedby="date-field--P0-0__hint date-field--P0-0__label-mask"
  aria-invalid="false"
  class="ds-c-field"
  id="date-field--P0-0"
  inputmode="numeric"
  name="ds-date-field"
  type="text"
/>
`;

exports[`DateField renders without picker 1`] = `
<input
  aria-describedby="static-id__hint static-id__label-mask"
  aria-invalid="false"
  class="ds-c-field"
  id="static-id"
  inputmode="numeric"
  name="ds-date-field"
  type="text"
/>
`;
